### ISSUE TITLE
GridFS ensure indexes is done without yielding

### ISSUE THREAD

**Issue by touilleMan** [twisted/txmongo#164]
https://github.com/twisted/txmongo/blob/master/txmongo/_gridfs/__init__.py#L55

When instanciating a `GridFS` object, an async call to mongo is done to create the chunks collection indexes.
The trouble is this call is never waited so this can lead to errors if the database is disconnected at the wrong time. This is especially true for tests where a lot of conections/disconnections are done in a short time (as a matter of fact I'm surprised this error didn't show up on travis continuous integration...):

```
(py35-twtrunk)➜  txmongo git:(issue-160) ✗ ./.tox/py35-twtrunk/bin/trial  tests.test_objects.TestGridFsObjects.test_GridFsObjects
tests.test_objects
  TestGridFsObjects
    test_GridFsObjects ...                                              [ERROR]

===============================================================================
[ERROR]
Traceback (most recent call last):
  File "/home/emmanuel/projects/txmongo/.tox/py35-twtrunk/lib/python3.5/site-packages/twisted/internet/defer.py", line 1126, in _inlineCallbacks
    result = result.throwExceptionIntoGenerator(g)
  File "/home/emmanuel/projects/txmongo/.tox/py35-twtrunk/lib/python3.5/site-packages/twisted/python/failure.py", line 389, in throwExceptionIntoGenerator
    return g.throw(self.type, self.value, self.tb)
  File "/home/emmanuel/projects/txmongo/txmongo/collection.py", line 704, in create_index
    yield self._database.system.indexes.insert(index, safe=True, **kwargs)
  File "/home/emmanuel/projects/txmongo/.tox/py35-twtrunk/lib/python3.5/site-packages/twisted/internet/defer.py", line 1126, in _inlineCallbacks
    result = result.throwExceptionIntoGenerator(g)
  File "/home/emmanuel/projects/txmongo/.tox/py35-twtrunk/lib/python3.5/site-packages/twisted/python/failure.py", line 389, in throwExceptionIntoGenerator
    return g.throw(self.type, self.value, self.tb)
  File "/home/emmanuel/projects/txmongo/txmongo/utils/__init__.py", line 27, in _timeout
    raw_result = yield raw_d
  File "/home/emmanuel/projects/txmongo/.tox/py35-twtrunk/lib/python3.5/site-packages/twisted/internet/defer.py", line 1126, in _inlineCallbacks
    result = result.throwExceptionIntoGenerator(g)
  File "/home/emmanuel/projects/txmongo/.tox/py35-twtrunk/lib/python3.5/site-packages/twisted/python/failure.py", line 389, in throwExceptionIntoGenerator
    return g.throw(self.type, self.value, self.tb)
  File "/home/emmanuel/projects/txmongo/txmongo/collection.py", line 357, in insert
    yield proto.get_last_error(str(self._database), **write_concern.document)
  File "/home/emmanuel/projects/txmongo/.tox/py35-twtrunk/lib/python3.5/site-packages/twisted/internet/defer.py", line 1126, in _inlineCallbacks
    result = result.throwExceptionIntoGenerator(g)
  File "/home/emmanuel/projects/txmongo/.tox/py35-twtrunk/lib/python3.5/site-packages/twisted/python/failure.py", line 389, in throwExceptionIntoGenerator
    return g.throw(self.type, self.value, self.tb)
  File "/home/emmanuel/projects/txmongo/txmongo/protocol.py", line 391, in get_last_error
    reply = yield self.send_QUERY(query)
pymongo.errors.AutoReconnect: TxMongo lost connection to MongoDB.

tests.test_objects.TestGridFsObjects.test_GridFsObjects
-------------------------------------------------------------------------------
Ran 1 tests in 0.016s

FAILED (errors=1)
```

One solution I see to solve this would be to store a defer on the ensure_indexes call in the GridFS

```python
>>> gfs = GridFS(db)
>>> # playing with gridfs
>>> yield gfs.ensure_indexes_defer
>>> db.disconnect()
```

Another smaller concern is on the indexes themselves, according to [mongodb's documentation](https://docs.mongodb.org/manual/core/gridfs/#gridfs-indexes):
> GridFS uses indexes on each of the chunks and files collections for efficiency. Drivers that conform to the GridFS specification automatically create these indexes for convenience. You can also create any additional indexes as desired to suit your application’s needs.

So I think txmongo should create indexes for the `fs.files` collection as well

**Comment by IlyaSkriblovsky** [comment]
Sounds like a reasonable solution. But I might be called it `indexes_created` or something.

**Comment by IlyaSkriblovsky** [comment]
@IlyaSkriblovsky And it must be a method (not attribute or property) that returns a new instance of Deferred on each call in order subscribers to not interfere with each other.

**Comment by IlyaSkriblovsky** [comment]
Closed by #167
